lib/testutils/src/proptest.rs

```diff
@@ -0,0 +1,322 @@
+use std::fmt::Debug;
+
+use itertools::Either;
+use itertools::Itertools as _;
+use jj_lib::repo_path::RepoPath;
+use jj_lib::repo_path::RepoPathBuf;
+use jj_lib::repo_path::RepoPathComponentBuf;
+use proptest::collection::btree_map;
+use proptest::collection::vec;
+use proptest::prelude::*;
+use proptest::sample::select;
+use proptest_state_machine::ReferenceStateMachine;
+
+#[derive(Debug, Clone)]
+pub struct RepoRefState {
+    root: Tree,
+}
+
+#[derive(Debug, Clone)]
+pub enum File {
+    RegularFile { contents: String, executable: bool },
+}
+
+#[derive(Debug, Clone)]
+pub enum Transition {
+    /// Create a file with the given contents at `path`.
+    ///
+    /// Parent directories are created as necessary, existing files or
+    /// directories are replaced.
+    CreateFile {
+        path: RepoPathBuf,
+        contents: String,
+        executable: bool,
+    },
+
+    /// Delete the file at `path`.
+    ///
+    /// Emptied parent directories are cleaned up (expect repo root).
+    DeleteFile { path: RepoPathBuf },
+}
+
+#[derive(Clone, Default)]
+struct Tree {
+    nodes: BTreeMap<RepoPathComponentBuf, Node>,
+}
+
+#[derive(Clone)]
+enum Node {
+    File(File),
+    Directory(Tree),
+}
+
+impl Debug for Tree {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.nodes
+            .iter()
+            .fold(&mut f.debug_struct("Directory"), |list, (name, node)| {
+                list.field(name.as_internal_str(), node)
+            })
+            .finish()
+    }
+}
+
+impl Debug for Node {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Node::File(file) => file.fmt(f),
+            Node::Directory(tree) => tree.fmt(f),
+        }
+    }
+}
+
+impl Default for Node {
+    fn default() -> Self {
+        Node::Directory(Tree::default())
+    }
+}
+
+impl RepoRefState {
+    pub fn files(&self) -> impl IntoIterator<Item = (RepoPathBuf, &File)> + '_ {
+        self.root.files(RepoPath::root())
+    }
+}
+
+impl Tree {
+    fn files(&self, prefix: &RepoPath) -> Vec<(RepoPathBuf, &File)> {
+        self.nodes
+            .iter()
+            .filter_map(move |(name, node)| match node {
+                Node::File(file) => Some((prefix.join(name), file)),
+                Node::Directory(_) => None,
+            })
+            .chain(
+                self.nodes
+                    .iter()
+                    .filter_map(move |(name, node)| match node {
+                        Node::File(_) => None,
+                        Node::Directory(tree) => Some((prefix.join(name), tree)),
+                    })
+                    .flat_map(|(path, tree)| tree.files(&path)),
+            )
+            .collect()
+    }
+}
+
+fn arb_tree() -> impl Strategy<Value = Tree> {
+    btree_map(arb_path_component(), arb_node(), 1..8).prop_map(|nodes| Tree { nodes })
+}
+
+fn arb_path_component() -> impl Strategy<Value = RepoPathComponentBuf> {
```

Comment:
suggestion (non-blocking): I'd recommend just generating `String`s or `PathBuf`s, and only converting them to `RepoPathComponentBuf`s right before feeding the values into the jj API, just as a matter of convenience, given that `RepoPathComponentBuf` isn't something we're specifically testing.